### DISCUSSION TITLE
Conversion from json to template type when I do not know the type beforehand.

### DISCUSSION THREAD

**Discussion by carul17** [nlohmann/json#3819] [Q&A] [answered]
I have to convert json files to a cpp class and I do not know if the json file will contain an int, a string, a vector of ints, a vector of maps, etc. This is the class:

```
class Game{
string name;

//ElemtnSptr is defined below the class ListElement given below
ElementSptr constants; //these are all lists
ElementSptr variables;
ElementSptr rules;
}
```

My questions is how do I go about converting the json file to this complex structure as I have no idea where to start. This is because ElementSptr is a type of ListElement. So far, I am only used to converting json to variables that have the same name. Here I want to convert the lists constants, variables and rules to a ListElement because these lists can contain other lists and I do not know the sublists names beforehand. So I have the ListElement class which can store other ListElements.

I have this setup:

```
class ListElement{
Type type;
//contains some virtual methods 
};

// I need to store recursive multi-type lists
typedef std::shared_ptr<ListElement> ElementSptr; 

typedef std::vector<std::shared_ptr<ListElement>> ElementVector; 

typedef std::map<std::string, std::shared_ptr<ListElement>> ElementMap; 

template <typename T> 
class Element : public ListElement{
T data;

Element(int data)              : _data(data) { type = Type::INT; }
    Element(std::string data)      : _data(data) { type = Type::STRING; }
    Element(ElementVector data)    : _data(data) { type = Type::VECTOR; }
    Element(ElementMap data)       : _data(data) { type = Type::MAP; }
   
//implementations of the virtual methods 
};
```

_data can be a an ElementSptr or ElementVector or ElementMap or a basic type. Further, ElementVector or ElementMap contains ListElements which means they can contain other ElementVectors or ElementMaps and so on. This supports the type of json data I have which contains lists containing lists containing lists and so on. But how do I go about the conversion?
If this is too complex, if someone could just tell me how to convert json data to a template type that would be good too.

**Comment by falbrechtskirchinger** [comment, marked as answer]
Give this a shot:
```cpp
void from_json(const json &j, ElementSptr &e) {
	if(j.is_string()) e = std::make_shared<Element<std::string>>(j.get<std::string>());
	else if(j.is_array()) e = std::make_shared<Element<ElementVector>>(j.get<ElementVector>());
	else if(j.is_object()) e = std::make_shared<Element<ElementMap>>(j.get<ElementMap>());
}
```

The default conversions for `std::vector` and `std::map` should do the rest.

**Reply by carul17** [reply]
I get the error:
```
error: no matching member function for call to 'get'
        else if(j.is_array()) e = std::make_shared<Element<ElementVector>>(j.get<ElementVector>());

```

**Reply by carul17** [reply]
Nevermind, I put this inside the element class and it was giving error. I move it outside and it is compiling. Will follow up with the results.

**Reply by carul17** [reply]
Ok, thank you! It is working. It is compiling and running.  I want to print it out to check if it's getting the right data, could you please tell me how I can write the to_json method for this?

**Reply by falbrechtskirchinger** [reply]
Here's a rough template.

```cpp
void to_json(json &j, const ElementSptr &e) {
	switch(e->type) {
		case Type::STRING:
			j = // assign string data in e
			break;
		case Type::VECTOR:
			j = // assign vector data in e
			break;
		// etc. ...
	}
}
```

**Reply by carul17** [reply]
Thank you very much! This has been extremely helpful.

**Reply by carul17** [reply]
Hi, for the string type, j = e should suffice right? But for vector and map, Im not really sure what to assign j. I tried use j = json::object(e) for the map and j = json::array(e) for the vector but I get a segmentation fault.
